cmds/msr/msr_linux.go

```diff
@@ -0,0 +1,62 @@
+// Copyright 2012-2017 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// This file contains support functions for io for Linux.
+package main
+
+import (
+	"encoding/binary"
+	"fmt"
+	"log"
+	"os"
+	"path/filepath"
+)
+
+func msrList(n string) []string {
+	m, err := filepath.Glob(filepath.Join("/dev/cpu", n, "msr"))
+	if err != nil {
+		log.Fatal(err)
+	}
+	return m
+}
+
+func doio(dev string, addr uint32, f func(*os.File) error) error {
+	port, err := os.OpenFile(string(dev), os.O_RDWR, 0)
+	if err != nil {
+		return err
+	}
+	defer port.Close()
+	_, err = port.Seek(int64(addr), 0)
+	if err != nil {
+		return fmt.Errorf("in: bad address %v: %v", addr, err)
+	}
+	return f(port)
+}
+
+// msr io rules are a bit weird. If you specify a single core, you probably want to know
+// if it does not exist. If you specify a big range, it's probably ok if some don't exist.
+// The question of error return therefore gets a bit tricky. We opt for only returning
+// and error if end was a value other than -1, -1 meaning 'all cores'
+func rdmsr(m []string, addr uint32) ([]uint64, []error) {
+	var (
+		regs = make([]uint64, len(m))
+		errs = make([]error, len(m))
+	)
+	for i := range m {
+		errs[i] = doio(m[i], addr, func(port *os.File) error {
```

Comment:
actually, making it parallel would almost certainly make it worse, since for each open you'd add the startup overhead of a process. This is one case where simple serialization is almost certainly best.